backend/httpclient/provider.go

```diff
@@ -0,0 +1,190 @@
+package httpclient
+
+import (
+	"crypto/tls"
+	"net/http"
+)
+
+// ProviderOptions options that will be used as default if not specified
+// in Options provided to Provider.New, Provider.GetTransport and
+// Provider.GetTLSConfig.
+type ProviderOptions struct {
+	// Timeouts timeout/connection related options.
+	Timeout *TimeoutOptions
+
+	TLS *TLSOptions
+
+	// Middlewares optionally provide additional middlewares.
+	Middlewares []Middleware
+
+	// ConfigureMiddleware optionally provide a ConfigureMiddlewareFunc
+	// to modify the middlewares chain.
+	ConfigureMiddleware ConfigureMiddlewareFunc
+
+	// ConfigureClient optionally provide a ConfigureClientFunc
+	// to modify the created http.Client.
+	ConfigureClient ConfigureClientFunc
+
+	// ConfigureTransport optionally provide a ConfigureTransportFunc
+	// to modify the created http.Client.
+	ConfigureTransport ConfigureTransportFunc
+
+	// ConfigureTLSConfig optionally provide a ConfigureTLSConfigFunc
+	// to modify the created http.Client.
+	ConfigureTLSConfig ConfigureTLSConfigFunc
+}
+
+// Provider the default HTTP client provider implementation.
+type Provider struct {
+	Opts ProviderOptions
+}
+
+// NewProvider creates a new HTTP client provider.
+// Optionally provide ProviderOptions options that will be used as default if
+// not specified in Options argument to Provider.New, Provider.GetTransport and
+// Provider.GetTLSConfig.
+// If no middlewares are provided in opts the DefaultMiddlewares() will be used. If you
+// provide middlewares you have to manually add the DefaultMiddlewares() for it to be
+// enabled.
+// Note: Middlewares will be executed in the same order as provided.
+func NewProvider(opts ...ProviderOptions) *Provider {
```

Comment:
No it doesn't have to, but with that you can reuse same function for calling it without args `NewProvider()` or with args `NewProvider(ProviderOptions{])`. At first I had `func NewProvider(*ProviderOptions)` that required `NewProvider(nil)` to call it without options so this got a bit cleaner. I can change it if you want to two functions `func NewProvider() *Provider` and `func NewProviderWithOptions(opts ProviderOptions) *Provider`. WDYT?

Reply:
Ah okay that makes sense - didn't think of that. I think since this pattern is used in a few places and keeps the boilerplate to a minimum, let's just keep it as is. We can only really validate how this should look when using it from the client side and I think this is a good first step

Reply:
This may be a signal that we may want sth like this instead:
```
func NewProvider(opts ...ProviderOption) *Provider
```
I.e. use functional option pattern here and pass separate options instead of many `ProviderOptions` structs.

Or use explicit:

```
func NewProvider(opts ProviderOptions) *Provider
```

Is there a bad thing to only have the last API?

Having a variadic argument here that breaks when passing 2 `ProviderOptions` seems less obvious from user perspective.

Reply:
> Is there a bad thing to only have the last API?

Not really, but it's convenient if you would like to use it without specifying any options. `NewProvider()` vs `NewProvider(ProviderOptions{})`.

> Having a variadic argument here that breaks when passing 2 ProviderOptions seems less obvious from user perspective.

It's a common pattern in many Go libraries to support optional arguments. I wouldn't say it break if 2 or more ProviderOptions are provided, but only the first one will be used.

So it sounds like having `func NewProvider() *Provider and func NewProviderWithOptions(opts ProviderOptions) *Provider is the way forward?

Reply:
But we still have same optional options with `New`, `GetTransport` and `GetTLSConfig` in http_client.go. Not very found duplicating all those with `WithOptions` functions. @FZambia WDYT?

Reply:
Here I really wish sth like Python `kwargs` or function pattern matching like in Java existed in a language.. I'd say if you like the current approach then let's at least write comment that function accepts zero or one `ProviderOptions` struct. Maybe adding panic in case of more than 1 option struct passed to help developer find a bug as soon as possible?

Reply:
Good point. Will do